### PR TITLE
[SPARK-42599][CONNECT][INFRA] Introduce `dev/connect-jvm-client-mima-check`  instead of `CompatibilitySuite`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
The main changes of this pr as follows:

- Refactor `CompatibilitySuite` as a new tool `CheckConnectJvmClientCompatibility`
- Add a new shell script `dev/connect-jvm-client-mima-check`, it will use `CheckConnectJvmClientCompatibility` to check the mima compatibility of `connect-jvm-client` module.
- Add `dev/connect-jvm-client-mima-check` to github task

### Why are the changes needed?
For fix test error report in `[VOTE] Release Apache Spark 3.4.0 (RC1)` mail list.

Testing `CompatibilitySuite` with maven requires some pre-work: 

```
build/mvn clean install -DskipTests -pl sql/core -am
build/mvn clean install -DskipTests -pl connector/connect/client/jvm -am 
```

So if we run `build/mvn package test` to test whole project as before, `CompatibilitySuite` will failed as follows:

```
CompatibilitySuite:
- compatibility MiMa tests *** FAILED ***
  java.lang.AssertionError: assertion failed: Failed to find the jar inside folder: /home/bjorn/spark-3.4.0/connector/connect/client/jvm/target
  at scala.Predef$.assert(Predef.scala:223)
  at org.apache.spark.sql.connect.client.util.IntegrationTestUtils$.findJar(IntegrationTestUtils.scala:67)
  at org.apache.spark.sql.connect.client.CompatibilitySuite.clientJar$lzycompute(CompatibilitySuite.scala:57)
  at org.apache.spark.sql.connect.client.CompatibilitySuite.clientJar(CompatibilitySuite.scala:53)
  at org.apache.spark.sql.connect.client.CompatibilitySuite.$anonfun$new$1(CompatibilitySuite.scala:69)
  at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
  at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
  at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
  at org.scalatest.Transformer.apply(Transformer.scala:22)
  at org.scalatest.Transformer.apply(Transformer.scala:20)
  ...
- compatibility API tests: Dataset *** FAILED ***
  java.lang.AssertionError: assertion failed: Failed to find the jar inside folder: /home/bjorn/spark-3.4.0/connector/connect/client/jvm/target
  at scala.Predef$.assert(Predef.scala:223)
  at org.apache.spark.sql.connect.client.util.IntegrationTestUtils$.findJar(IntegrationTestUtils.scala:67)
  at org.apache.spark.sql.connect.client.CompatibilitySuite.clientJar$lzycompute(CompatibilitySuite.scala:57)
  at org.apache.spark.sql.connect.client.CompatibilitySuite.clientJar(CompatibilitySuite.scala:53)
  at org.apache.spark.sql.connect.client.CompatibilitySuite.$anonfun$new$7(CompatibilitySuite.scala:110)
  at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
  at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
  at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
  at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
  at org.scalatest.Transformer.apply(Transformer.scala:22) 
```

So we need to fix this problem for developers.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GitHub Actions

### CODE REVIEW

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala [hunk 1]
```diff
@@ -190,30 +215,32 @@ class CompatibilitySuite extends ConnectFunSuite {
...
+    problems
   }
 
-  test("compatibility API tests: Dataset") {
-    val clientClassLoader: URLClassLoader = new URLClassLoader(Seq(clientJar.toURI.toURL).toArray)
-    val sqlClassLoader: URLClassLoader = new URLClassLoader(Seq(sqlJar.toURI.toURL).toArray)
-
-    val clientClass = clientClassLoader.loadClass("org.apache.spark.sql.Dataset")
-    val sqlClass = sqlClassLoader.loadClass("org.apache.spark.sql.Dataset")
+  private def checkDatasetApiCompatibility(clientJar: File, sqlJar: File): Seq[String] = {
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala:221]: Note: When `CheckConnectJvmClientCompatibility` in the `connect/jvm` module, this check will fail due to https://github.com/apache/spark/pull/39712#discussion_r1118448264

Comment (zhenlineo) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala:221]: Hi, I would recommend to use this PR. For this failing tests. Let's just delete this. I was planning to delete it after this PR get merged. The test case is already covered by mima test. Thanks so much. This change looks awesome.

**Comment by zhenlineo** [comment]
LGTM

**Review by zhenlineo — APPROVED** [review]

File: dev/connect-jvm-client-mima-check [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+if [[ -x "$JAVA_HOME/bin/java" ]]; then
+  JAVA_CMD="$JAVA_HOME/bin/java"
+else
+  JAVA_CMD=java
+fi
+
+rm -f .connect-mima-check-result
+
+echo "Build sql module ..."
+build/sbt sql/package
```
Comment (hvanhovell) [inline dev/connect-jvm-client-mima-check:38]: You can issue a single command here, that will be a bit faster:
```
build/sbt "sql/package;connect-client-jvm/assembly;connect-client-jvm/Test/package"
```

File: dev/connect-jvm-client-mima-check [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+echo "Build connect-client-jvm module ..."
+build/sbt connect-client-jvm/assembly
+echo "Build connect-client-jvm test jar ..."
+build/sbt connect-client-jvm/Test/package
+
+CONNECT_TEST_CLASSPATH="$(build/sbt -DcopyDependencies=false "export connect-client-jvm/Test/fullClasspath" | grep jar | tail -n1)"
+CONNECT_CLASSPATH="$(build/sbt -DcopyDependencies=false "export connect-client-jvm/fullClasspath" | grep jar | tail -n1)"
+SQL_CLASSPATH="$(build/sbt -DcopyDependencies=false "export sql/fullClasspath" | grep jar | tail -n1)"
```
Comment (hvanhovell) [inline dev/connect-jvm-client-mima-check:46]: There will be duplication of entries here. Is that an issue?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 4]
```diff
@@ -2739,7 +2739,7 @@ class Dataset[T] private[sql] (
     sparkSession.analyze(plan, proto.AnalyzePlanRequest.AnalyzeCase.SCHEMA)
   }
 
-  def collectResult(): SparkResult[T] = sparkSession.execute(plan, encoder)
+  private[sql] def collectResult(): SparkResult[T] = sparkSession.execute(plan, encoder)
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2742]: As mentioned in the other PR. Please do not change this.

**Comment by hvanhovell** [comment]
@LuciferYang can we merge this one instead of https://github.com/apache/spark/pull/40191.

**Comment by amaliujia** [comment]
Looks awesome overall!

**Comment by amaliujia** [comment]
How would we deal with `CompatibilitySuite`?

**Comment by amaliujia** [comment]
Ah I see now that we still need to update the excluding rules.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala [hunk 5]
```diff
@@ -185,30 +210,32 @@ class CompatibilitySuite extends ConnectFunSuite {
...
+
+    def methods(jar: File, className: String): Seq[String] = {
+      val classLoader: URLClassLoader = new URLClassLoader(Seq(jar.toURI.toURL).toArray)
```
Comment (zhenlineo) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala:219]: To be more safe about the class is loaded correctly. We can also use this class loader:
```
import org.apache.spark.util.ChildFirstURLClassLoader

val classLoader: URLClassLoader = new ChildFirstURLClassLoader(Seq(jar.toURI.toURL).toArray, this.getClass.getClassLoader)
```

Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala:219]: Thanks

File: dev/connect-jvm-client-mima-check (hunk 3 above)
Comment (LuciferYang) [reply dev/connect-jvm-client-mima-check:46]: Yes, there will be duplicate entries, but there is no problem for the time being,

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala (hunk 5 above)
Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala:219]: done

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 4 above)
Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2742]: `checkDatasetApiCompatibility('compatibility API tests: Dataset' in CompatibilitySuite)` will check find out the public Dataset api in `connect/client` module  instead of in `sql` module. `collectResult ` is such a case. There was no problem before because there was bad case in the test (classloader loaded the wrong class)

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2742]: > I guess it is because this is not an existing API?

Yes

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2742]: I think If we accept similar changes (add some new public apis to `connect/client` module), we may need to maintain a whitelist(in checkDatasetApiCompatibility) or do not check this again



checkDatasetApiCompatibility

**Comment by LuciferYang** [comment]
> How would we deal with `CompatibilitySuite`
> Ah I see now that we still need to update the excluding rules.

Yes, just check in a different way :)

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 4 above)
Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2742]: This is one of the only places where we will be deviating from the original API. I do intent to create the same class for the original API, but that is for later.

File: dev/connect-jvm-client-mima-check (hunk 2 above)
Comment (LuciferYang) [reply dev/connect-jvm-client-mima-check]: Thanks ~ fixed

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 4 above)
Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2742]: without this change, this check will print error message as follows:

```
ERROR: Dataset apis incompatible with the sql module include: 
org.apache.spark.sql.Dataset.collectResult
connect-client-jvm module mima check failed.
Exceptions to binary compatibility can be added in tools/CheckConnectJvmClientCompatibility.scala
```

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala (hunk 5 above)
Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala]: This is effective. I have removed the change of `Dataset.scala` and it can find out `org.apache.spark.sql.Dataset.collectResult` not exist in `sql` module

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 4 above)
Comment (amaliujia) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2742]: It comes to the question of do we pursue this check to verify that Scala client is a subset of the original API or intersect with it. Per my understanding, current implementation is to check the subset.

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2742]: > changed to check intersection

What does this mean? Is the intersection compatible? It seems that `checkMiMaCompatibility` has already done this?

Comment (amaliujia) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2742]: I will leave @hvanhovell to comment and see the right way to address this. 

To be clear I was thinking based on this
```
    // For now we simply check the new methods is a subset of the old methods.
    clientMethods.diff(sqlMethods)
  }
```

I thought the comment says that we only check if new methods are a subset of the existing methods otherwise it will fail. And you need to change the method otherwise it becomes not a subset.

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2742]: Do you mean we should change this to 

```
    clientMethods.filterNot(m => sqlMethods.contains(m))
```
?

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2742]:
```scala
    // For now we simply check the new methods is a subset of the old methods.
    clientMethods.diff(sqlMethods)
```

`the new methods is a subset of the old method`  -> `clientMethods.diff(sqlMethods)` should be empty?

For example:

```
scala> val clientMethods = Seq(1,2,3)
clientMethods: Seq[Int] = List(1, 2, 3)

scala> val sqlMethods = Seq(1,2,3,4,5)
sqlMethods: Seq[Int] = List(1, 2, 3, 4, 5)

scala> clientMethods.diff(sqlMethods)
res0: Seq[Int] = List()
```

`clientMethods` is a subset of `sqlMethods`.


```
scala> val clientMethods = Seq(0,1,2,3)
clientMethods: Seq[Int] = List(0, 1, 2, 3)

scala> val sqlMethods = Seq(1,2,3,4,5)
sqlMethods: Seq[Int] = List(1, 2, 3, 4, 5)

scala> clientMethods.diff(sqlMethods)
res0: Seq[Int] = List(0)
```

`clientMethods` is not a subset of `sqlMethods`.

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2742]: The original test code as follows. I think it is consistent with the current intention, am I right?

```
// For now we simply check the new methods is a subset of the old methods.
    newMethods
      .map(m => m.toString)
      .foreach(method => {
        assert(oldMethods.map(m => m.toString).contains(method))
      })
```

Comment (amaliujia) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2742]: Please feel free to ignore me and work @hvanhovell on this case. :)

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala (hunk 5 above)
Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala]: And I do manually check for the original test `compatibility API tests: Dataset`, it also became failed:

[failed.log](https://github.com/apache/spark/files/10867177/failed.log) (The error message is too large. I pasted it into the file)

**Comment by LuciferYang** [comment]
image

The new change of `org.apache.spark.sql.Dataset#plan` in SPARK-42631 is checked as incompatible.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 4 above)
Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2742]: @LuciferYang can we make an exception for the SparkResult class. Basically filter out the DataSet.collectResult method. That is all.

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2742]: let me add a white list to `checkDatasetApiCompatibility `

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala [hunk 6]
```diff
@@ -184,30 +218,36 @@ class CompatibilitySuite extends ConnectFunSuite {
...
+        .toSeq
+    }
 
-    val newMethods = clientClass.getMethods
-    val oldMethods = sqlClass.getMethods
+    val className = "org.apache.spark.sql.Dataset"
+    val clientMethods = methods(clientJar, className)
+    val sqlMethods = methods(sqlJar, className)
+    // Exclude some public methods that must be added through `exceptions`
+    val exceptionMethods =
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala:246]: @hvanhovell `exceptionMethods` has been added and the revert the change of `Dataset`. Can we simply fix it like this first?

Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala:246]: If I can refine this check, I will give followup later

Comment (hvanhovell) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala:246]: looks good

**Review by hvanhovell — APPROVED** [review]
LGTM

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala [hunk 7]
```diff
@@ -184,30 +218,36 @@ class CompatibilitySuite extends ConnectFunSuite {
...
-        s"\nComparing client jar: $clientJar\nand sql jar: $sqlJar\n" +
-          problems.map(p => p.description("client")).mkString("\n"))
-    }
+    problems
   }
 
-  test("compatibility API tests: Dataset") {
-    val clientClassLoader: URLClassLoader = new URLClassLoader(Seq(clientJar.toURI.toURL).toArray)
-    val sqlClassLoader: URLClassLoader = new URLClassLoader(Seq(sqlJar.toURI.toURL).toArray)
+  private def checkDatasetApiCompatibility(clientJar: File, sqlJar: File): Seq[String] = {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala:224]: @zhenlineo is this only supposed to check Dataset? what about the other classes?

**Comment by hvanhovell** [comment]
Alirght merging.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala (hunk 7 above)
Comment (zhenlineo) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala:224]: This test was initially added because we did not cover the Dataset in the mima check. 

If we found such test is useful and want to expand to more classes we can run the mima check the other way around to ensure that our new code does not introduce more unwanted methods.

**Comment by LuciferYang** [comment]
Thanks @hvanhovell @amaliujia @zhenlineo
